### DISCUSSION TITLE
Better Tags

### DISCUSSION THREAD

**Discussion by DarkArc** [standardnotes/forum#2694] [Suggestions]
I'd really like a way to express tag queries... I'm not sure if this is doable via the extension system or not.

For instance if I have the tags:
- A
- B
- C

I could write something like:

`A && B && !C`

To get all notes that have tags `A`, and `B`, but don't have the tag `C`. I'd strongly prefer this over creating `ABC!` permutations of tags, or `A-B` tags, as I think it greatly improves usability.

----

As a more practical example, maybe you have a `meeting` tag, and a `my-current-job` tag. Well, lots of things are presumably tagged with `my-current-job`, but maybe you want to find the meetings for `my-current-job`; `meeting && my-current-job`, is a lot easier/scales better than adding `my-current-job-meeting`, `my-old-job-meeting`, `my-old-old-job-meeting`.

Would also be awesome if you could save `meeting & my-current-job` as a custom view for quick access. If you change jobs, just change the search, or delete the view.

Hope this makes sense.

----

I wanted to look for the `folders` app to see if I could peace together the practicality of doing this. I couldn't find the sources for that one though under https://github.com/sn-extensions.

**Comment by JaspalSuri** [comment]
I appreciate the feedback and will pass it along to the team. 

> I wanted to look for the folders app to see if I could peace together the practicality of doing this. I couldn't find the sources for that one though under https://github.com/sn-extensions.

I'll get back to you with a link to the folders extension if it's available.

**Comment by arielsvg** [comment]
Hey @DarkArc, we have a dynamic view syntax that's documented in our help section: https://standardnotes.org/help/42/how-do-i-view-a-list-of-untagged-notes-and-create-other-dynamic-filters would this be what you're looking for?

**Comment by DarkArc** [comment]
@baptiste-grob kind of, it's close. I think there needs to be a way to do ad-hoc searching though. The grammar is a bit cryptic as well. I could see it being hard for non-programmers to understand.

Just having some way to advanced search tags (and optionally save that as a view) would be useful. I think personally what I want more is having better tag search, over having saved views. Saved views are just a nicety for some things. I've seen things like the following work reasonably well without introducing too much complexity:

```
All of these: #meeting, #important
Some of these: #ipo, #investors
None of these: #stock-options
```

This would find notes that have both `#meeting` and `#important`, that also have either `#ipo` or `#investors`, and exclude any that are tagged as `#stock-options`.

Some more examples:

I'm hungry
```
All of these: #food
Some of these: #take-out, #5-minute-meal
None of these: #frozen-meal
```

Recent evernote import
```
All of these: #evernote, #my-company
Some of these: #ideas, #todo
```

What could I get into today
```
Some of these: #ideas, #todo
```

There are a lot of potential use cases for something a little nicer WRT tag search :)

**Comment by DarkArc** [comment]
I had another idea about how this might be able to be integrated into standard notes...

You could move the plus for views lower, add a "filter button" (I used an eye):

![image](https://user-images.githubusercontent.com/778012/76811522-334c9b00-67c8-11ea-88fc-baa4e834c128.png)

You could then click on the eye to create a tag filter:
![image](https://user-images.githubusercontent.com/778012/76811741-0baa0280-67c9-11ea-8a49-692242c624f9.png)

Then perhaps that pops in a button which allows you to name and save the search as a view?
![image](https://user-images.githubusercontent.com/778012/76811781-2bd9c180-67c9-11ea-97aa-b1d2e1cdcc2f.png)

What I like about this, is it takes over the tag area, which already acts as a sort of filter when you click on a tag. I think this feels more natural/simpler than an advanced search, and it could naturally be combined with a normal search, similarly to how when you click on a tag, you filter your search by that tag.

Thoughts?

(Edit: Obviously rather rough mockups :) )

Edit 2: Perhaps after filling out the filters, a + could appear where the current tag `+` is that allows you to create a view?

**Comment by JaspalSuri** [comment]
Thanks for those working on those mockups to help illustrate what you're suggesting! 

[Here's something that the Drafts 5 app has.](https://imgur.com/a/gvSZQzy.png)

**Comment by DarkArc** [comment]
@JaspalSuri sorry for delayed reply. Something about the Drafts 5 screenshot you posted doesn't quite do it for me. Having the functionality alone would be awesome, but the way that's presented with all the icons and things, just feels... clunky/busy?

I don't have time to create a mockup for this... I think perhaps a better way to do what I was proposing up above might be an autocomplete similar to the "Quick Tags" extension. 

Ultimately, I think this is something that needs to integrate well with keyboard shortcuts down the line, and I'm not sure the Drafts 5 design would get you there. 

EDIT: It's worth noting there may be some conflicting interest here design wise as what's good for desktop may not be good for mobile and vice versa. A touch screen would likely feel better under Drafts 5, and worse under a keyboard/mouse setup. 

This would probably mean you'd want kind of a hybrid UI element. One where on desktop you get an autocompleting search, and on mobile you get a full screen searchable list displayed on the screen you can just "tap" things on.

I'm also not totally satisfied with my original mockup. I feel it puts too much pressure on the user to save the tag configuration as a view. Ideally there'd be a reset and save button of equal prominence.

**Comment by JaspalSuri** [comment]
Don't worry too much about the design part; we appreciate the feedback! :) 

I've taken note of the option to filter and combine tags and need for more granular search options, though!

**Comment by darkdragon-001** [comment]
I like the workflow how it is done in Gmail: Use the normal search bar with [special keywords/syntax](https://support.google.com/mail/answer/7190), extend the search bar to show fields like _All of these words_ with a small triangle. When you are happy with the results, there is another button to create a filter. This way, the UI doesn't become more cluttered (only the small triangle), but is useful for normal users with the expanded menu as well as power users with keywords.

**Comment by fineseasalt** [comment]
Thank you for your feedback and ideas 🙏 @DarkArc @darkdragon-001
